File: ios/Classes/AgoraRtcEnginePlugin.swift
import Flutter
import UIKit
import AgoraRtcKit

public class AgoraRtcEnginePlugin: NSObject, FlutterPlugin {
  private let engineHolder = AgoraRtcEngineViewHolder()
  private let eventStreamHandler = AgoraRtcEventStreamHandler()
  private lazy var actions = AgoraRtcEngineActions(
    engineHolder: engineHolder,
    emitter: eventStreamHandler
  )

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "plugins.flutter.io/agora_rtc",
      binaryMessenger: registrar.messenger()
    )
    let eventChannel = FlutterEventChannel(
      name: "plugins.flutter.io/agora_rtc/events",
      binaryMessenger: registrar.messenger()
    )
    let instance = AgoraRtcEnginePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    eventChannel.setStreamHandler(instance.eventStreamHandler)
    registrar.register(
      AgoraRtcVideoViewFactory(engineHolder: instance.engineHolder, isLocal: true),
      withId: "plugins.flutter.io/agora_rtc/local_view"
    )
    registrar.register(
      AgoraRtcVideoViewFactory(engineHolder: instance.engineHolder, isLocal: false),
      withId: "plugins.flutter.io/agora_rtc/remote_view"
    )
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "createEngine":
      actions.handleCreateEngine(call, result: result)
    case "destroyEngine":
      actions.handleDestroyEngine(result: result)
    case "joinChannel":
      actions.handleJoinChannel(call, result: result)
    case "leaveChannel":
      actions.handleLeaveChannel(result: result)
    case "updateChannelMediaOptions":
      actions.handleUpdateChannelMediaOptions(call, result: result)
    case "setClientRole":
      actions.handleSetClientRole(call, result: result)
    case "muteAllRemoteAudioStreams":
      actions.handleMuteAllRemoteAudioStreams(call, result: result)
    case "setChannelProfile":
      actions.handleSetChannelProfile(call, result: result)
    case "renewToken":
      actions.handleRenewToken(call, result: result)
    case "muteAllRemoteVideoStreams":
      actions.handleMuteAllRemoteVideoStreams(call, result: result)
    case "muteRemoteAudioStream":
      actions.handleMuteRemoteAudioStream(call, result: result)
    case "muteRemoteVideoStream":
      actions.handleMuteRemoteVideoStream(call, result: result)
    case "muteLocalAudioStream":
      actions.handleMuteLocalAudioStream(call, result: result)
    case "muteLocalVideoStream":
      actions.handleMuteLocalVideoStream(call, result: result)
    case "setRemoteVideoStreamType":
      actions.handleSetRemoteVideoStreamType(call, result: result)
    case "enableVideo":
      actions.handleEnableVideo(call, result: result)
    case "enableLocalVideo":
      actions.handleEnableLocalVideo(call, result: result)
    case "startPreview":
      actions.handleStartPreview(result: result)
    case "stopPreview":
      actions.handleStopPreview(result: result)
    case "takeSnapshot":
      actions.handleTakeSnapshot(call, result: result)
    case "startRecording":
      actions.handleStartRecording(call, result: result)
    case "stopRecording":
      actions.handleStopRecording(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

File: android/src/main/kotlin/com/changteng/agora_rtc/AgoraRtcEnginePlugin.kt
package com.changteng.agora_rtc

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.StandardMessageCodec

/** AgoraRtcEnginePlugin */
class AgoraRtcEnginePlugin : FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private var applicationContext: Context? = null
  private val engineHolder = AgoraRtcEngineHolder()
  private val eventStreamHandler = AgoraRtcEventStreamHandler()
  private val actions = AgoraRtcEngineActions(engineHolder, eventStreamHandler)

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    applicationContext = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "plugins.flutter.io/agora_rtc")
    channel.setMethodCallHandler(this)
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "plugins.flutter.io/agora_rtc/events")
    eventChannel.setStreamHandler(eventStreamHandler)
    flutterPluginBinding.platformViewRegistry.registerViewFactory(
      "plugins.flutter.io/agora_rtc/local_view",
      AgoraLocalVideoViewFactory(engineHolder, StandardMessageCodec.INSTANCE)
    )
    flutterPluginBinding.platformViewRegistry.registerViewFactory(
      "plugins.flutter.io/agora_rtc/remote_view",
      AgoraRemoteVideoViewFactory(engineHolder, StandardMessageCodec.INSTANCE)
    )
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "createEngine" -> actions.handleCreateEngine(call, result, applicationContext)
      "destroyEngine" -> actions.handleDestroyEngine(result)
      "joinChannel" -> actions.handleJoinChannel(call, result)
      "leaveChannel" -> actions.handleLeaveChannel(result)
      "updateChannelMediaOptions" -> actions.handleUpdateChannelMediaOptions(call, result)
      "setClientRole" -> actions.handleSetClientRole(call, result)
      "muteAllRemoteAudioStreams" -> actions.handleMuteAllRemoteAudioStreams(call, result)
      "setChannelProfile" -> actions.handleSetChannelProfile(call, result)
      "renewToken" -> actions.handleRenewToken(call, result)
      "muteAllRemoteVideoStreams" -> actions.handleMuteAllRemoteVideoStreams(call, result)
      "muteRemoteAudioStream" -> actions.handleMuteRemoteAudioStream(call, result)
      "muteRemoteVideoStream" -> actions.handleMuteRemoteVideoStream(call, result)
      "muteLocalAudioStream" -> actions.handleMuteLocalAudioStream(call, result)
      "muteLocalVideoStream" -> actions.handleMuteLocalVideoStream(call, result)
      "setRemoteVideoStreamType" -> actions.handleSetRemoteVideoStreamType(call, result)
      "enableVideo" -> actions.handleEnableVideo(call, result)
      "enableLocalVideo" -> actions.handleEnableLocalVideo(call, result)
      "startPreview" -> actions.handleStartPreview(call, result)
      "stopPreview" -> actions.handleStopPreview(call, result)
      "takeSnapshot" -> actions.handleTakeSnapshot(call, result)
      "startRecording" -> actions.handleStartRecording(call, result)
      "stopRecording" -> actions.handleStopRecording(result)
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
    applicationContext = null
    engineHolder.clear()
    actions.clearRuntimeState()
  }
}
